include/types/register.hpp

```diff
@@ -1,94 +1,45 @@
-    }
-}
-}  // namespace detail
-
-/*
- * @brief The register tile is comprised of a 2-dimensional array of data
- *        elements.
- * @tparam Element_: the data type of the elements.
- * @tparam TemporalExec_: the atomic memory access instruction is executed in a
- *         temporal fashion along two dimensions. The first element of
- *         TemporalExec_ is the number of stripes along the outer dimension,
- *         and the second element is the number of stripes along the inner
- *         dimension.
- * @tparam ElemTileLayout_: the data tile layout for executing a single atomic
- *         memory access instruction at a time.
- */
-template <typename Element, typename TemporalExec_, typename ElemTileLayout_,
-          typename Base = traits::TraitsBase<Element>>
-class RegTile : public Base {
+template <typename Element_, typename Layout_>
+class RegTile {
   public:
-    using DType = Element;
-    using TemporalExec = TemporalExec_;
-    using ElemTileLayout = ElemTileLayout_;
-
-    // how many 128-bit are used per access
-    constexpr static int kRegCountPerAccess =
-        Base::kAccessInBits / 8 / sizeof(uint32_t);
+    using DType = Element_;
+    using Layout = Layout_;
 
-    constexpr static int kRows =
-        dim_size<0, ElemTileLayout> * dim_size<0, TemporalExec>;
-    constexpr static int kCols =
-        dim_size<1, ElemTileLayout> * dim_size<1, TemporalExec>;
+    static constexpr int kNumel = tl::get_numel<Layout>;
+    static constexpr int kRows = tl::num_rows<Layout>;
+    static constexpr int kCols = tl::num_cols<Layout>;
 
-    // how many times the atomic memory access instruction is executed
-    constexpr static int kExecCount = get_numel<TemporalExec>;
-
-    constexpr static int kElemInsNumel = get_numel<ElemTileLayout>;
-    constexpr static int kNumel = kRows * kCols;
-
-    DEVICE RegTile() { memset((void*)data_, 0, sizeof(data_)); }
-
-    DEVICE DType* operator[](int2 idx) { return &data_[idx.x][idx.y]; }
-
-    DEVICE const DType* operator[](int2 idx) const {
-        return &data_[idx.x][idx.y];
+    DEVICE RegTile() : layout_(Layout{}) {
+        memset((void*)data_, 0, sizeof(data_));
     }
 
     DEVICE DType* mutable_data() { return (DType*)data_; }
 
     DEVICE const DType* data() const { return (DType*)data_; }
 
-    template <typename T>
-    DEVICE static void dump_value(const T* data, int numel) {
-        detail::dump_value<T>(data, numel);
-    };
+    // for write access
+    DEVICE DType& operator()(int x, int y) { return data_[layout_(x, y)]; }
+
+    // for read access
+    DEVICE const DType& operator()(int x, int y) const {
+        return data_[layout_(x, y)];
+    }
+
+    DEVICE void dump_value() {
+        print_tile(static_cast<DType*>(data_), layout_);
+    }
 
   private:
-    // register tile data is implemented as an 2D array of size kRows x kCols
-    DType data_[kRows][kCols];
+    DType data_[kNumel];
+    Layout layout_;
```

Comment:
Good, it seems more flexible than ThunderKittens.

Reply:
I have a feeling that, CuTe's `Layout` is a good expressive tool. I am still trying to make the best use of it. 😊